library/Freckle/App/Http/Cache/Memcached.hs

```diff
@@ -0,0 +1,144 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedRecordDot #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Freckle.App.Http.Cache.Memcached
+  ( memcachedHttpCacheSettings
+  , memcachedHttpCodec
+  , memcachedHttpCache
+  ) where
+
+import Freckle.App.Prelude
+
+import Blammo.Logging (MonadLogger, logWarnNS)
+import Codec.Serialise (Serialise (..), deserialiseOrFail, serialise)
+import qualified Data.ByteString.Lazy as BSL
+import Data.CaseInsensitive (CI)
+import qualified Data.CaseInsensitive as CI
+import Data.String (IsString (..))
+import Database.Memcache.Types (Value)
+import Freckle.App.Http.Cache
+import Freckle.App.Memcached
+import qualified Freckle.App.Memcached.Client as Memcached
+import Freckle.App.OpenTelemetry (MonadTracer)
+import Network.HTTP.Client (Cookie, CookieJar, Request)
+import qualified Network.HTTP.Client.Internal as HTTP
+import Network.HTTP.Types.Header (ResponseHeaders)
+import Network.HTTP.Types.Status (Status)
+import Network.HTTP.Types.Version (HttpVersion (..))
+
+memcachedHttpCacheSettings
+  :: ( MonadUnliftIO m
+     , MonadLogger m
+     , MonadTracer m
+     , MonadReader env m
+     , HasMemcachedClient env
+     , Serialise body
+     )
+  => CacheTTL
+  -- ^ Default TTL, used when @max-age@ is not present
+  -> HttpCacheSettings m body Value
+memcachedHttpCacheSettings defaultTTL =
+  HttpCacheSettings
+    { shared = False
+    , cacheable = const True
+    , defaultTTL
+    , getCurrentTime = liftIO getCurrentTime
+    , logWarn = logWarnNS "http.cache" . fromString
+    , codec = memcachedHttpCodec
+    , cache = memcachedHttpCache
+    }
+
+memcachedHttpCodec :: Serialise body => HttpCacheCodec body Value
+memcachedHttpCodec =
+  HttpCacheCodec
+    { serialise = BSL.toStrict . serialise . fromResponse
+    , deserialise = \req ->
+        bimap show (toResponse req)
+          . deserialiseOrFail
+          . BSL.fromStrict
+    }
+
+memcachedHttpCache
+  :: ( MonadUnliftIO m
+     , MonadTracer m
+     , MonadReader env m
+     , HasMemcachedClient env
+     )
+  => HttpCache m Value
+memcachedHttpCache =
+  HttpCache
+    { get = try . Memcached.get
+    , set = \k v -> try $ Memcached.set k v 0
+    , evict = try . Memcached.delete
+    }
+
+data CResponse body = CResponse
+  { cresponseStatus :: Status
+  , cresponseVersion :: HttpVersion
+  , cresponseHeaders :: ResponseHeaders
+  , cresponseBody :: body
+  , cresponseCookieJar :: CookieJar
```

Comment:
Why are cookies here at all? That does not seem like something you'd care to cache. I am only able to imagine why this could be bad and not why it could be good.

Reply:
Barring any other trade-offs, I was just trying to cache as much as I could to keep a cached response and a real response as identical as I can. I feel like that's still the right thing to do.

Why would it be bad?

Reply:
It seems like what would be most bad is if we ever actually used those cookies, creating some complex consequences of trying to cache requests within stateful sessions? I don't know. Vague unknown-unknown concern.

Only concrete concern is - why slightly inflate the cache size with information that is redundant to the response headers?

Reply:
> that is redundant to the response headers

Yeah, that's my mistake. For some reason, I was picturing something like `Set-Cookie` as information that is in the Cookie Jar, and so retaining it would be necessary to be a faithful representation of the response. But you're right, obviously that's all in the headers and the Jar is a client state thing added by http-client that has nothing to do with information coming from the server in this response.

I will remove it.